File: framework/languages/dotnet/src/Zlink.Framework/Runtime/Locations/ZLinkLocationRegistration.cs
namespace Zlink.Framework.Runtime.Locations;

/// <summary>
/// Raw location store registration collected by the options builder. The
/// five store roles must share one physical store. Registration is either
/// one store instance or the in-memory store.
/// </summary>
internal sealed class ZLinkLocationRegistration
{
    private ZLinkInMemoryLocationStore? _inMemoryStore;

    public bool UseInMemoryStores { get; set; }

    /// <summary>One physical store instance providing every store role
    /// (draft 20.2). Optional contracts on the same instance (change stamp,
    /// watch) are registered automatically.</summary>
    public IZLinkLocationStore? StoreInstance { get; set; }

    public IZLinkRelocationStore? RelocationStoreInstance { get; set; }

    public ZLinkLocationOptions Options { get; } = new();

    public bool Enabled => UseInMemoryStores || StoreInstance is not null;

    internal IZLinkLocationStore? ResolveStore()
    {
        if (StoreInstance is not null) return StoreInstance;
        if (!UseInMemoryStores) return null;
        return _inMemoryStore ??= new ZLinkInMemoryLocationStore();
    }
}

File: framework/languages/java/zlink-framework-core/src/main/java/systems/zlink/framework/runtime/locations/ZLinkLocationRegistration.java
package systems.zlink.framework.runtime.locations;

import systems.zlink.framework.locations.ZLinkLocationOptions;
import systems.zlink.framework.locations.ZLinkLocationStore;

public final class ZLinkLocationRegistration {
    private final ZLinkLocationOptions options = new ZLinkLocationOptions();
    private boolean useInMemoryStores;
    private ZLinkLocationStore storeInstance;

    public ZLinkLocationOptions options() {
        return options;
    }

    public boolean useInMemoryStores() {
        return useInMemoryStores;
    }

    public void enableInMemoryStores() {
        this.useInMemoryStores = true;
    }

    public ZLinkLocationStore storeInstance() {
        return storeInstance;
    }

    public void setStoreInstance(ZLinkLocationStore storeInstance) {
        this.storeInstance = storeInstance;
    }

    public boolean enabled() {
        return useInMemoryStores || storeInstance != null;
    }
}
